arbnode/espresso_batcher_addr_monitor.go

```diff
@@ -0,0 +1,154 @@
+package arbnode
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+const (
+	eventKey = "espresso-batcher-addr"
+)
+
+// `BatchPosterSet` event
+type BatcherAddrEvent struct {
+	l1Height  uint64         `koanf:"l1-height"`
+	addr      common.Address `koanf:"addr"`
+	isBatcher bool           `koanf:"is-batcher"`
+}
+
+type BatcherAddrMonitorInterface interface {
+	AddBatchPosterSetEvent(l1 uint64, addr common.Address, isBatcher bool) error
+	// Returns the valid batcher addresses up to the given L1 height
+	GetValidAddresses(l1 uint64) []common.Address
+	// The latest L1 height that has been seen by the monitor. This implies
+	// that all events up to this height have been processed.
+	SetL1Height(l1 uint64)
+}
+
+var _ BatcherAddrMonitorInterface = (*BatcherAddrMonitor)(nil)
+
+type BatcherAddrMonitor struct {
+	l1Height          uint64
+	lastEventL1Height uint64
+	cached            bool
+	// Cache for the latest valid addresses.
+	// Since batcher address changes are infrequent and callers typically
+	// process HotShot blocks sequentially, caching improves performance.
+	cachedAddresses []common.Address
+
+	events        []BatcherAddrEvent
+	db            ethdb.Database
+	initAddresses []common.Address
+}
+
+func NewBatcherAddrMonitor(
+	addr []common.Address,
+	db ethdb.Database,
+) *BatcherAddrMonitor {
+	return &BatcherAddrMonitor{
+		initAddresses: addr,
+		db:            db,
+	}
+}
+
+func (b *BatcherAddrMonitor) AddBatchPosterSetEvent(l1 uint64, addr common.Address, isBatcher bool) error {
+	event := BatcherAddrEvent{
```

Comment:
I dont think this event actually matches the event in the contract?

Reply:
https://github.com/EspressoSystems/nitro-contracts/blob/develop/src/bridge/SequencerInbox.sol#L965 - this is the correct event

Reply:
Isn't this one?
https://github.com/EspressoSystems/nitro-contracts/blob/c6f2078eb04acb4a0146172e7d4d2f4b3c2a3d0e/src/bridge/SequencerInbox.sol#L964

I just added a l1 height here